cpp/src/compaction/apply_stencil.cu

```diff
@@ -0,0 +1,925 @@
+{
+    size_t   scratch_space_size = 0;
+    // Note: We assume this can't fail when only a scratch space size calculation is requested.
+    // The cub documentation indicate that is the case - but only rather implicitly.
+    cub::DeviceScan::ExclusiveSum<T*, T*>(nullptr, scratch_space_size, nullptr, nullptr, num_elements);
+    return scratch_space_size;
+}
+
+gdf_size_type make_aligned_size(gdf_size_type size)
+{
+    enum : gdf_size_type { alignment_quantum = 256 };
+    return cudf::util::round_up_to_power_of_2<gdf_size_type>(size, alignment_quantum);
+}
+
+
+// This class should be merely a templated lambda within gpu_apply_stencil,
+// but support for those is not available before C++17, while this code is
+// targetting C++14
+template <
+    typename Size,
+    unsigned ThreadsPerBlock,
+    unsigned InputElementsPerThread
+>
+struct apply_stencil_helper {
+    template <typename T>
+    gdf_error operator()(
+        void*                 __restrict__  selected_data,
+        gdf_bool*             __restrict__  unpacked_selected_data_validity,
+        gdf_size_type*        __restrict__  selected_data_null_count,
+        const void*           __restrict__  input_data,
+        const bit_container*  __restrict__  input_data_validity,
+        const gdf_bool*       __restrict__  stencil,
+        const bit_container*  __restrict__  stencil_validity,
+        Size                                input_length,
+        const Size*           __restrict__  block_pass_counts_prefix_sums,
+        cudaStream_t                        stream
+    )
+    {
+        auto selection_grid_config { cudf::util::form_naive_1d_grid(input_length, ThreadsPerBlock, InputElementsPerThread) };
+
+        auto kernel = (stencil_validity == nullptr) ?
+            kernels::stably_apply_stencil_with_block_pass_counts<Size, T, ThreadsPerBlock, InputElementsPerThread, false> :
+            kernels::stably_apply_stencil_with_block_pass_counts<Size, T, ThreadsPerBlock, InputElementsPerThread, true>;
+
+        kernel
+            <<<
+                selection_grid_config.num_blocks,
+                ThreadsPerBlock,
+                cudf::util::cuda::no_dynamic_shared_memory,
+                stream
+            >>>
+            (
+                static_cast<T*>(selected_data),
+                unpacked_selected_data_validity,
+                selected_data_null_count,
+                static_cast<const T*>(input_data),
+                input_data_validity,
+                stencil,
+                stencil_validity,
+                block_pass_counts_prefix_sums,
+                input_length
+            );
+
+        CUDA_TRY (cudaGetLastError() );
+        return GDF_SUCCESS;
+    }
+};
+
+} // namespace detail
+
+
+/*
+ * @note: all of `stencil->data`, `column->valid`, `stencil->valid` and `output->valid`
+ * must have their start and end addresses of be multiples of 4, i.e. interpretable as
+ * `bit_container *`
+ */
+gdf_error gpu_apply_stencil(
```

Comment:
Haven't done a proper review pass yet, but did notice that this needs it's Doxymentation updated in `functions.h`.

Reply:
Also, how about renaming to `gdf_apply_boolean_mask`? @harrism how do you feel about that naming?

Reply:
@jrhemstad : I don't mind that, but does it really help? I mean, a "boolean mask" is also n t nullable.. Do you want it to be `apply_nullable_boolean_mask`? .

Reply:
> I don't mind that, but does it really help?

Yes, "applying a stencil" has [a very well understood meaning](https://en.wikipedia.org/wiki/Stencil_code) that this function is most definitely _not_ doing and it is misleading. 

> Do you want it to be apply_nullable_boolean_mask `apply_nullable_boolean_mask`

No. The fact that you're accepting a `gdf_column` indicates that the column could have null values. There's no reason to add the verbosity to the name. Furthermore, your naming would suggest there should be an additional API `apply_nonnullable_boolean_mask`, but there isn't and there shouldn't be, because the internals of `apply_boolean_mask` should be able to handle both code paths.

Reply:
@jrhemstad : I can accept your reasoning. So be it. @harrism , @felipeblazing , @williamBlazing : Are you ok with `apply_boolean_mask`?

Reply:
works for me

Reply:
@jrhemstad : 
> Did notice that this needs it's Doxymentation updated in functions.h.

It does. That wasn't what this PR was about, but have now added a decent Doxygen comment. Waiting for your proper review pass. Actually, this is a month-old PR, so I want to nudge you to move faster on this.

Reply:
Boy Scout's rule: Leave it better than you found it :smile: 

I apologize that the review has lagged, we're on the verge of a new release that has many high priority issues and features we're trying to resolve and include for the 0.6 release.